src/modem.rs

```diff
@@ -0,0 +1,335 @@
+use at_commands::{builder::CommandBuilder, parser::CommandParser};
+use atat::{self, AtatCmd};
+use core::{borrow::BorrowMut, ffi::c_void, marker::PhantomData};
+use esp_idf_hal::{
+    delay::{TickType, BLOCK},
+    io::EspIOError,
+    uart::{UartDriver, UartTxDriver},
+};
+
+use crate::{
+    eventloop::EspSystemEventLoop,
+    handle::RawHandle,
+    netif::{EspNetif, EspNetifDriver, NetifStack},
+    sys::*,
+};
+
+pub struct EspModem<'d, T>
+where
+    T: BorrowMut<UartDriver<'d>>,
+{
+    serial: T,
+
+    _d: PhantomData<&'d ()>,
+}
+
+impl<'d, T> EspModem<'d, T>
+where
+    T: BorrowMut<UartDriver<'d>> + Send,
+{
+    pub fn new(serial: T) -> Self {
+        Self {
+            serial,
+            _d: PhantomData,
+        }
+    }
+
+    pub fn send_cmd<CMD: AtatCmd>(&mut self, cmd: &CMD) -> Result<CMD::Response, atat::Error> {
+        let mut buff = [0u8; 64];
+        // flush the channel
+        // self.serial
+        //     .borrow_mut()
+        //     .clear_rx()
+        //     .map_err(|_err| atat::Error::Write)?;
+
+        // write the command to the uart
+        let len = cmd.write(&mut buff);
+        log::info!("about to write {:?}", &buff[..len]);
+        self.serial
+            .borrow_mut()
+            .write(&buff[..len])
+            .map_err(|_err| atat::Error::Write)?;
+
+        // now read the uart to get the response
+
+        let len = self
+            .serial
+            .borrow_mut()
+            .read(&mut buff, TickType::new_millis(1000).ticks())
+            .map_err(|_err| atat::Error::Read)?;
+        log::info!("got response {:?}", &buff[..len]);
+        cmd.parse(Ok(&buff[..len]))
+    }
+
+    pub fn setup_data_mode(&mut self, sysloop: EspSystemEventLoop) -> Result<(), EspError> {
+        self.reset()?;
+        //disable echo
+        self.set_echo(false)?;
+
+        // get iccid
+        self.get_iccid()?;
+
+        // check pdp network reg
+        self.read_gprs_registration_status()?;
+
+        //configure apn
+        self.set_pdp_context()?;
+
+        // start ppp
+        self.set_data_mode()?;
+
+        // now in ppp mode.
+        let netif = EspNetif::new(NetifStack::Ppp)?;
+        // subscribe to user event
+        esp!(unsafe {
+            esp_event_handler_register(
+                IP_EVENT,
+                ESP_EVENT_ANY_ID as _,
+                Some(Self::raw_on_ip_event),
+                netif.handle() as *mut core::ffi::c_void,
+            )
+        })?;
+        esp!(unsafe {
+            esp_event_handler_register(
+                NETIF_PPP_STATUS,
+                ESP_EVENT_ANY_ID as _,
+                Some(Self::raw_on_ppp_changed),
+                netif.handle() as *mut core::ffi::c_void,
+            )
+        })?;
+        let (mut tx, rx) = self.serial.borrow_mut().split();
+        let driver = unsafe {
```

Comment:
This line - and the `loop {}` below is probably the one thing in `EspModem/SimModem` which is specific to the underlying platform (i.e. ESP-IDF, `embassy-net-ppp` and Linux). We can probably consider isolating this code in a separate function which is `#[cfg(target_os = "espidf")]`. Or even better - we can introduce a trait - `PppRunner` or something that the `SimModem` expects on constructor and which trait is responsible for connecting the `Read/Write` IO to the underlying PPP stack once the negotiation with the modem is complete.

Something like (in blocking-IO terms, for now):
```rust
trait PppRunner {
    /// Hand the read/write pipes of the modem to the underlying platform PPP stack once 
    /// the negotiation with the modem with AT commands is over and it is time to switch to PPP data mode
    fn run<R, W>(&mut self, read: R, write: W) -> Result<(), R::Error>
    where 
        Self: Sized,
        R: embedded_io::Read,
        W: embedded_io::Write<Error = R::Error>;
}
```

... and the implementation for e.g. ESP IDF would be something like:
```rust
struct EspPppRunner(());

impl EspPppRunner {
    pub const fn new() -> Self {
        Self(())
    }

    pub fn run<R, W>(read: R, write: W) -> Result<Self, EspError> 
    where 
        R: embedded_io::Read,
        W: embedded_io::Write<Error = R::Error>,
    {
        let netif = EspNetif::new(NetifStack::Ppp)?;
        let driver = EspNetifDriver::new_ppp(move |data| esp_idf_hal::io::Write::write_all(write, data).map_err(|e| e.0))?;

        let mut buff = [0u8; 64];

        loop {
            let len = rx.read(&mut buff, BLOCK)?;
            if len > 0 {
                driver.rx(&buff[..len])?;
            }
        }

        // TODO: We should watch out for PPP frames that signify end of PPP communication and exit the loop    
    }
}

impl PppRunner for EspPppRunner {
    fn run<R, W>(&mut self, read: R, write: W) -> Result<(), R::Error> 
    where 
        R: embedded_io::Read,
        W: embedded_io::Write<Error = R::Error>,
    {
        EspPppRunner::run(self, read, write)
    }
}
```

Note also, that what we can also do is **not even deal** with the PPP stack in your `SimModem` implementation, and this way it would become pure Rust. 

Basically, rather than renaming `setup_data_mode` to    `run` - as I suggested above - we can leave it named `setup_data_mode` or shorten to `negotiate` or something, and then just document that once `negotiate` is over, the user is supposed to establish the native PPP connection of the `read/write` pipes to the underlying platform, using platform-specific means (whatever they are) without even bothering to introduce a `PppRunner` trait to your `SimModem`, because - after all - the main value of your `SimModem` would be in the pre-PPP "AT commands" phase where - ideally and over time - it would gain support for other various modems besides `SIM7600`.

Reply:
I like the idea to change this to `negotiate`. let 'run' become a user concern.